pyinteraph/path_analysis.py

```diff
@@ -0,0 +1,481 @@
+                path = list(nx.algorithms.shortest_paths.generic.all_shortest_paths(\
+                                G = graph, \
+                                source = node1, \
+                                target = node2))
+                for p in path:
+                    # Check that path is not longer than the maximum allowed length
+                    if len(p) <= maxl:
+                        paths.append(p)
+            except nx.NetworkXNoPath:
+                pass
+                # If no path is found log info
+                log.warning(f"No path found between {node1} and {node2}")
+    return paths
+
+def get_all_simple_paths(graph, source, target, maxl):
+    """Find all simple paths between all combinations of source and
+    target.
+    """
+
+    # Get all combinations
+    combinations = itertools.product(source, target)
+    # Get all simple paths
+    paths = []
+    for node1, node2 in combinations:
+        # Get all simple paths for each combination of source and target
+        path = list(nx.algorithms.simple_paths.all_simple_paths(\
+                        G = graph, \
+                        source = node1, \
+                        target = node2, \
+                        cutoff= maxl))
+        # Only add paths to output if they exist
+        for p in path:
+            if len(p) > 0:
+                paths.append(p)
+    return paths
+
+def sort_paths(graph, paths, sort_by):
+    """Takes in a list of paths and sorts them."""
+    
+    # Get source and target
+    source = [p[0] for p in paths]
+    target = [p[-1] for p in paths]
+    # Calculate length of path
+    lengths = [len(p) for p in paths]
+    #print(lengths)
+    # Calculate weights of path
+    weights = \
+    [[graph[p[i]][p[i+1]]["weight"] for i in range(len(p)-1)] for p in paths]
+    sum_weights = [np.sum(w) for w in weights]
+    avg_weights = [np.mean(w) for w in weights]
+    # Sort paths
+    paths = zip(paths, source, target, lengths, sum_weights, avg_weights)
+    if sort_by == "length":
+        key = lambda x: x[3]
+        reverse = False
+    elif sort_by == "cumulative_weight":
+        key = lambda x: x[4]
+        reverse = True
+    elif sort_by == "average_weight":
+        key = lambda x: x[5]
+        reverse = True
+    sorted_paths = sorted(paths, key = key, reverse = reverse)
+    #print(sorted_paths)
+    return sorted_paths
+
+def get_combinations(res_id, res_space):
+    """ Takes in a list of residue identifiers and returns all pairs of
+    residues that are at least res_space apart if they are on the same
+    chain.
+    """
+    
+    # Get all index combinations
+    combinations = itertools.combinations(range(len(res_id)), 2)
+    # Get all residue combinations if they are res_space distance apart
+    # Or they are on different chains
+    combinations = [(res_id[idx1], res_id[idx2]) for idx1, idx2 in combinations \
+        if abs(idx1 - idx2) > res_space or res_id[idx1][0] != res_id[idx2][0]]
```

Comment:
It is clearer if you move this part of the list at the same level of indentation as the portion in the previous line (i.e. move the `if` right below the starting `[(`).